### PR TITLE
New parallelization scheme for subgraph isomorphism

### PR DESCRIPTION
**Pull request by averbukh**
# Description
New parallelization scheme for subgraph isomorphism

Changes proposed in this pull request:
- mutex added
- atomic operations added
- global stack added and used for work stealing in subgraph isomorphism threads

### CODE REVIEW

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp [hunk 1]
```diff
@@ -298,6 +364,7 @@ vertex_stack<Cpu>::vertex_stack(const std::uint64_t max_states_size, inner_alloc
     stack_size = max_states_size;
     stack_data = allocator_.allocate<std::uint64_t>(stack_size);
     ptop = stack_data;
+    bottom_ = stack_data;
 }
 
 template <typename Cpu>
```
Comment (orazve) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp:370]: remark: looks like we need to delete this constructor

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp [hunk 2]
```diff
@@ -365,17 +444,136 @@ graph_status vertex_stack<Cpu>::increase_stack_size() {
...
+
+    return false;
+}
+
+template <typename Cpu>
+void global_stack<Cpu>::pop(dfs_stack<Cpu>& s) {
+    ONEDAL_ASSERT(s.empty());
+    const dal::detail::scoped_lock lock(mutex_);
+    if (!empty()) {
+        // const auto& v = data_.top();
```
Comment (orazve) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp:487]: remove comment

File: cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp [hunk 3]
```diff
@@ -365,17 +444,136 @@ graph_status vertex_stack<Cpu>::increase_stack_size() {
...
+    ONEDAL_IVDEP
+    for (auto dest = new_bottom, src = bottom_; dest != new_top;) {
+        *(dest++) = *(src++);
+    }
+
+    clear();
+
+    bottom_ = new_bottom;
+    top_ = new_top;
+    capacity_ = new_capacity;
```
Comment (orazve) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp:574]: reset ptop?

Comment (averbukh) [reply cpp/oneapi/dal/algo/subgraph_isomorphism/backend/cpu/stack.hpp:574]: what is ptop? top_ is assigned in line 573.

**Review by orazve — APPROVED** [review]

**Pull request merged**